### DISCUSSION TITLE
conversion of const reference to value in input parameters

### DISCUSSION THREAD

**Discussion by ad8e** [carbon-language/carbon-lang#1908] [Language design]
A [recent blog post about Carbon's calling convention](https://www.foonathan.net/2022/07/carbon-calling-convention/) says that a const ref parameter (const int&) can be invisibly converted to a copy, if it's cheaper.

1. Is that the correct interpretation? The [Carbon design document](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design#parameters) says "This binding will be implemented using a pointer, unless it is legal to copy and copying is cheaper." Does the word "legal" just mean a copyable type? It is unclear here. The original blog post calls this an "as-if" transformation, which would require checking if the parameter is changed by another reference before being able to convert to a copy.
2. If the simple interpretation is correct, the design should also state that a const reference parameter can assume the object is unchanging for the relevant lifetime. This is the case when converting a const reference parameter to a copy parameter. It would enable a performance increase: loading a value, calling `puts()`, and loading the value again, would make it valid to elide one of the two loads. Otherwise, there would be a difference in semantics between small types (which can be converted to copies) and large types.

Converting a const reference to a copy, and being able to assume that reference is to an unchanging object, have the same meaning. So if one is true, the other should be true too.

**Comment by jonmeow** [comment]
Yes, that's a correct interpretation. There's already discussion of this at https://github.com/carbon-language/carbon-lang/issues/1830 if you're interested. But you're not finding a design because this isn't an area that has been fully designed out.
